File: scheduler/internal/app/repositories/models/History.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type History struct {
	ID         uuid.UUID  `gorm:"type:uuid;primaryKey"`
	UserID     uuid.UUID  `gorm:"type:uuid;not null"`
	DrugID     uuid.UUID  `gorm:"type:uuid;not null"`
	GroupID    *uuid.UUID `gorm:"type:uuid"`
	Status     string     `gorm:"type:text;check:status IN ('taken', 'missed', 'skipped')"`
	TakenAt    *time.Time `gorm:"type:timestamp"`
	NotifiedAt time.Time  `gorm:"type:timestamp;not null"`
	Count      int        `gorm:"type:int;default:0;not null"`

	User AppUser `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE;"`
}

func (History) TableName() string {
	return `history`
}

func (h *History) BeforeCreate(tx *gorm.DB) (err error) {
	h.ID = uuid.New()
	return nil
}

File: drugapi/src/main/java/com/mahidol/drugapi/history/models/entities/History.java
package com.mahidol.drugapi.history.models.entities;

import com.mahidol.drugapi.history.models.types.TakenStatus;
import com.mahidol.drugapi.history.models.types.TakenStatusConvertor;
import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Data
@Accessors(chain = true)
@NoArgsConstructor
@Table(name = "history")
public class History {
    @Id
    private UUID id;

    @Column(name = "user_id", columnDefinition = "UUID")
    private UUID userId;

    @Column(name = "drug_id", columnDefinition = "UUID")
    private UUID drugId;

    @Column(name = "group_id", columnDefinition = "UUID")
    private UUID groupId;

    @Convert(converter = TakenStatusConvertor.class)
    @Column(name = "status", columnDefinition = "TEXT")
    private TakenStatus status;

    @Column(name = "taken_at", columnDefinition = "TIMESTAMP")
    private LocalDateTime takenAt;

    @Column(name = "notified_at", columnDefinition = "TIMESTAMP")
    private LocalDateTime notifiedAt;

    @Column(name = "count", columnDefinition = "INT")
    private Integer count;
}
